bin/kafka-run-class.sh

```diff
@@ -212,7 +212,7 @@ fi
 
 # JMX port to use
 if [  $JMX_PORT ]; then
-  KAFKA_JMX_OPTS="$KAFKA_JMX_OPTS -Dcom.sun.management.jmxremote.port=$JMX_PORT "
+  KAFKA_JMX_OPTS="-Dcom.sun.management.jmxremote.rmi.port=$JMX_PORT $KAFKA_JMX_OPTS -Dcom.sun.management.jmxremote.port=$JMX_PORT "
```

Comment:
Is there an official spec on how the `java` command should parse system properties on the command line when multiple values are provided? If there isn't, we should make sure not to specify the same value twice:
```suggestion
  KAFKA_JMX_OPTS="$KAFKA_JMX_OPTS -Dcom.sun.management.jmxremote.port=$JMX_PORT "
  if ! echo "$KAFKA_JMX_OPTS" | grep -qF -- '-Dcom.sun.management.jmxremote.rmi.port=' ; then
    # If unset, set the RMI port to address issues with monitoring Kafka running in containers
    KAFKA_JMX_OPTS="$KAFKA_JMX_OPTS -Dcom.sun.management.jmxremote.rmi.port=$JMX_PORT"
  fi
```

Reply:
@C0urante it seems like there is not good answer to that, but for simple properties (like `-D`), the "winner" is the right-most value in the argument list on most JDKs (HotSpot, OpenJ9).

Reply:
@reta I'd prefer to rely on officially-documented behavior. Do you see an issue with the suggested approach?

Reply:
> @reta I'd prefer to rely on officially-documented behavior. Do you see an issue with the suggested approach?

I bet :-) Here is the thing (which is often exploited), assume that `$KAFKA_JMX_OPTS` comes with `-Dcom.sun.management.jmxremote.port=xxxx` (from environment or deployment let say) and user wants to override `com.sun.management.jmxremote.port` with different port. How she should approach it?

Reply:
> @C0urante it seems like there is not good answer to that, but for simple properties (like `-D`), the "winner" is the right-most value in the argument list on most JDKs (HotSpot, OpenJ9).

I think so too, so I think it should be a straightforward implementation with a minimum amount of code

Reply:
> > @reta I'd prefer to rely on officially-documented behavior. Do you see an issue with the suggested approach?
> 
> I bet :-) Here is the thing (which is often exploited), assume that `$KAFKA_JMX_OPTS` comes with `-Dcom.sun.management.jmxremote.port=xxxx` (from environment or deployment let say) and user wants to override `com.sun.management.jmxremote.port` with different port. How she should approach it?

In this case, instead of specifying the 'JMX_PORT' environment variable, the user might just use 'KAFKA_JMX_OPTS', which would not cause problems

Reply:
Cool, so no, @C0urante I don't see any issues with this approach

Reply:
Thanks @reta. To your point--in general, if users want to rely on this undocumented behavior of prioritizing the rightmost value for a system property, they can also use the `$KAFKA_OPTS` variable to declare overrides in case they do not want to alter, e.g., `$KAFKA_JMX_OPTS`.

@a364176773 I understand the desire to simplify implementation, but I'm also trying to satisfy two important principles of bug-squashing:
1. Limit changes only to the areas that are known to be broken (which in this case, is when the `$JMX_PORT` variable is specified but not the `com.sun.management.jmxremote.rmi.port` system property)
2. Avoid depending on platform-specific behavior that is not guaranteed by a specification, docs, etc. (which in this case is the prioritization of the rightmost value for a system property on the command line when invoking `java`)

I think the cost of an extra `grep` invocation is a reasonable tradeoff to pay in order to satisfy these tenets, and would like to see that before merging.

Another committer may disagree and feel comfortable merging these changes as-are, but given my limited familiarity with RMI, I'm hesitant to merge this on my own without making absolutely sure that it will have no unintended consequences.

Reply:
@C0urante I understand your concerns very well and think that adopting your ideas is the best of both worlds, so why not? I will make a change as soon as possible